packages/contract/lib/contract/constructorMethods.js

```diff
@@ -154,44 +154,59 @@ module.exports = Contract => ({
     delete this.network.address;
   },
 
+  // accepts 4 input formats
+  //  - (<name>, <address>)
+  //  - (<contractType>) - must have a deployed instance with an address
+  //  - (<contractInstance>)
+  //  - ({ <libName>: <address>, <libName2>: <address2>, ... })
   link(name, address) {
-    // Case: Contract.link(instance)
-    if (typeof name === "function") {
-      const contract = name;
-
-      if (contract.isDeployed() === false) {
-        throw new Error("Cannot link contract without an address.");
-      }
+    switch (typeof name) {
+      case "string":
+        // Case: Contract.link(<libraryName>, <address>)
+        if (this._json.networks[this.network_id] == null) {
+          this._json.networks[this.network_id] = {
+            events: {},
+            links: {}
+          };
+        }
+
+        this.network.links[name] = address;
+        return;
+      case "function":
+        // Case: Contract.link(<contractType>)
+        const contract = name;
 
-      this.link(contract.contractName, contract.address);
+        if (contract.isDeployed() === false) {
+          throw new Error("Cannot link contract without an address.");
+        }
 
-      // Merge events so this contract knows about library's events
-      Object.keys(contract.events).forEach(topic => {
-        this.network.events[topic] = contract.events[topic];
-      });
+        this.link(contract.contractName, contract.address);
 
-      return;
-    }
-
-    // Case: Contract.link({<libraryName>: <address>, ... })
-    if (typeof name === "object") {
-      const obj = name;
-      Object.keys(obj).forEach(name => {
-        const a = obj[name];
-        this.link(name, a);
-      });
-      return;
-    }
-
-    // Case: Contract.link(<libraryName>, <address>)
-    if (this._json.networks[this.network_id] == null) {
-      this._json.networks[this.network_id] = {
-        events: {},
-        links: {}
-      };
+        // Merge events so this contract knows about library's events
+        Object.keys(contract.events).forEach(topic => {
+          this.network.events[topic] = contract.events[topic];
+        });
+        return;
+      case "object":
+        // 2 Cases:
+        //   - Contract.link({<libraryName>: <address>, ... })
+        //   - Contract.link(<instance>)
+        const obj = name;
+        if (obj.constructor && obj.constructor.contractName && obj.address) {
+          // obj is a Truffle contract instance
+          this.link(obj.constructor.contractName, obj.address);
+        } else {
+          // obj is of the form { <libraryName>: <address>, ... }
+          Object.keys(obj).forEach(name => {
+            const a = obj[name];
+            this.link(name, a);
+          });
+        }
+        return;
+      default:
+        const message = "Input to the link method is in the incorrect format.";
```

Comment:
Can we specify the expected formats in the error message?

Reply:
I did consider this but it might be a little bit long-winded. However it is probably better to err on the side of being more helpful! I'll go add more detail.